rz-svd/rz-svd.c

```diff
@@ -0,0 +1,295 @@
+#include <rz_cmd.h>
+#include <rz_cons.h>
+#include <rz_core.h>
+#include <rz_lib.h>
+#include <rz_types.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <yxml.h>
+
+#define XMLBUFSIZE 4096
+
+typedef struct ta_iter {
+	uint8_t yxml_buf[4096];
+	yxml_t x;
+
+	const char *start;
+	const char *ptr;
+	const char *end;
+
+	char regname[50];
+	char groupName[50];
+	char bitoffset[10];
+	char bitwidth[20];
+	char description[50];
+} ta_iter;
+
+static void strcpytrunc(char *dst, const char *src, size_t dstsize) {
+	size_t to_copy = strlen(src);
+	if (to_copy >= dstsize)
+		to_copy = dstsize - 1;
+	memcpy(dst, src, to_copy);
+	dst[to_copy] = '\0';
+}
+
+static const RzCmdDescArg cmd_svd_args[] = {
+	{
+		.name = "svd file",
+		.type = RZ_CMD_ARG_TYPE_STRING,
+
+	},
+	{ 0 },
+};
+
+static const RzCmdDescHelp svd_usage = {
+	.summary = "SVD Plugin for Rizin",
+	.args = cmd_svd_args,
+};
+
+static inline int ta_iter_done(ta_iter *ta) {
+	return *ta->ptr == 0 || ta->ptr >= ta->end;
+}
+
+static ta_iter *ta_iter_next(ta_iter *ta) {
+	yxml_ret_t r = YXML_OK;
+	yxml_t ta_x;
+
+	int level;
+	char *cur, *tmp;
+	char value[2048];
+	const char *ta_start;
+	enum { REGNAME,
+		BITWIDTH,
+		BITOFFSET,
+		DESCRIPTION } elem_type;
+
+	cur = value;
+	value[0] = 0;
+
+	if (!ta->groupName[0]) {
+		while (!ta_iter_done(ta) &&
+			(yxml_parse(&ta->x, *ta->ptr) != YXML_ELEMSTART ||
+				strcasecmp(ta->x.elem, "peripherals"))) {
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+		ta_start = ta->ptr;
+		ta_x = ta->x;
+
+		level = 0;
+		while (!ta_iter_done(ta) && !ta->groupName[0]) {
+			switch ((r = yxml_parse(&ta->x, *ta->ptr))) {
+			case YXML_ELEMSTART:
+				level += 1;
+				if (level == 1 && strcasecmp(ta->x.elem, "groupName") == 0) {
+					cur = value;
+					*cur = 0;
+				}
+				break;
+
+			case YXML_ELEMEND:
+				level -= 1;
+				if (level < 0) {
+					return ta_iter_next(ta);
+				} else if (level == 1 && cur) {
+					strcpytrunc(ta->groupName, value, sizeof(ta->groupName));
+					cur = NULL;
+					ta->ptr = ta_start;
+					ta->x = ta_x;
+				}
+				break;
+
+			case YXML_CONTENT:
+				if (!cur || level != 2) {
+					break;
+				}
+				tmp = ta->x.data;
+				while (*tmp && cur < value + sizeof(value)) {
+					*cur++ = *tmp++;
+				}
+				if (cur >= value + sizeof(value)) {
+					cur = NULL;
+				} else {
+					*cur = 0;
+				}
+				break;
+			default:
+				break;
+			}
+			ta->ptr++;
+		}
+		if (ta_iter_done(ta))
+			return NULL;
+	}
+	assert(ta->groupName[0]);
+
+	level = 0;
+	while (!ta_iter_done(ta)) {
+		r = yxml_parse(&ta->x, *ta->ptr);
+
+		if (r == YXML_ELEMSTART) {
+			level += 1;
+			if (level == 3 &&
+				strcasecmp(ta->x.elem, "register") == 0) {
+				break;
+			}
+
+		} else if (r == YXML_ELEMEND) {
+			level -= 1;
+			if (level < 0) {
+				ta->groupName[0] = 0;
+				return ta_iter_next(ta);
+			}
+		}
+		ta->ptr++;
+	}
+	if (ta_iter_done(ta)) {
+		return NULL;
+	}
+
+	assert(r == YXML_ELEMSTART);
```

Comment:
I am not sure about plugins, but since we are using RzUtil anyway, probably better to use `rz_return_if_fail()` instead of the `assert()`.

Reply:
Sure, I'm just confused on what exactly to pass to `rz_return_if_fail()` so that the logic is preserved.

Reply:
The same condition should work just fine.

Reply:
That throws up some warnings, as the function is non-void.
```
In file included from /usr/local/include/librz/rz_vector.h:5,
                 from /usr/local/include/librz/rz_cons.h:21,
                 from /usr/local/include/librz/rz_diff.h:6,
                 from /usr/local/include/librz/rz_util.h:8,
                 from /usr/local/include/librz/rz_io.h:10,
                 from /usr/local/include/librz/rz_analysis.h:14,
                 from ../rz-svd.c:1:
../rz-svd.c: In function ‘ta_iter_next’:
/usr/local/include/librz/rz_util/rz_assert.h:100:4: warning: ‘return’ with no value, in function returning non-void [-Wreturn-type]
  100 |    return; \
      |    ^~~~~~
../rz-svd.c:156:2: note: in expansion of macro ‘rz_return_if_fail’
  156 |  rz_return_if_fail(r == YXML_ELEMSTART);
      |  ^~~~~~~~~~~~~~~~~
../rz-svd.c:62:17: note: declared here
   62 | static ta_iter *ta_iter_next(ta_iter *ta) {
      |                 ^~~~~~~~~~~~
```

Reply:
Yes, because in case of the function returning something you need to use `rz_return_val_if_fail()` instead.

Reply:
Replaced with an if/return statement. 👍🏽